toolkit/bridge/primitives/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+#![deny(missing_docs)]
+
+extern crate alloc;
+
+use alloc::vec::*;
+use parity_scale_codec::{Decode, DecodeWithMemTracking, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use serde::{Deserialize, Serialize};
+use sp_inherents::*;
+
+#[cfg(feature = "std")]
+use {sp_api::ApiExt, sp_api::ProvideRuntimeApi, sp_runtime::traits::Block as BlockT};
+
+/// Smart contract hashes and addresses used by the token bridge
+#[derive(
+	Default,
+	Debug,
+	Clone,
+	PartialEq,
+	Eq,
+	TypeInfo,
+	Encode,
+	Decode,
+	DecodeWithMemTracking,
+	MaxEncodedLen,
+	Serialize,
+	Deserialize,
+)]
+pub struct MainChainScripts {}
+
+/// Represents a signle bridge transfer, corresponding to a single UTXO on Cardano.
+#[derive(
+	Clone, Debug, Encode, Decode, DecodeWithMemTracking, TypeInfo, PartialEq, Eq, MaxEncodedLen,
+)]
+pub enum BridgeTransferV1<RecipientAddress> {
+	/// Token transfer initiated by a user on Cardano
+	UserTransfer {
+		/// Amount of tokens tranfered
+		token_amount: u128,
+		/// Transfer recipient on the Partner Chain
+		recipient: RecipientAddress,
+	},
+	/// Token transfer carrying treasury funds being moved from Cardano to the Partner Chain
+	TreasuryTransfer {
+		/// Amount of tokens tranfered
+		token_amount: u128,
+	},
+	/// Invalid transfer coming from a malformed UTXO on Cardano that does not contain data
+	/// that can be correctly interpreted. These transfers can either be ignored and considered
+	/// lost or recovered through some custom mechanism.
+	InvalidTransfer {
+		/// Amount of tokens tranfered
+		token_amount: u128,
+		/// ID of the UTXO containing an invalid transfer
+		utxo_id: sidechain_domain::UtxoId,
+	},
+}
+
+/// Structure representing all token bridge transfers incoming from Cardano that are to be
+/// handled in one Partner Chain block.
+#[derive(Clone, Debug, Encode, Decode, TypeInfo, PartialEq)]
+pub struct TokenBridgeTransfersV1<RecipientAddress> {
+	/// List of token transfers registered for a single block
+	pub transfers: Vec<BridgeTransferV1<RecipientAddress>>,
+}
+
+/// Inherent identifier used by the Partner Chains token bridge
+pub const INHERENT_IDENTIFIER: InherentIdentifier = *b"pctokbrg";
+
+/// Error type returned by failing calls of the bridge inherent
+#[derive(Debug, Encode, Decode, thiserror::Error, PartialEq)]
+pub enum InherentError {
+	/// Inherent was not produced when expected
+	#[error("Bridge inherent data was present but an inherent was not provided")]
+	InherentRequired,
+	/// Inherent data is not correctly encoded
+	#[error("Bridge inherent data is not correctly encoded")]
+	InvalidInherentData,
+	/// Inherent produced when inherent data not present
+	#[error("Bridge inherent produced when no data present")]
+	InherentNotExpected,
+	/// Inherent produced does not match inherent data
+	#[error("Inherent produced does not match inherent data")]
+	IncorrectInherent,
+	/// Too many transfers to be processed in a single block
+	#[error("Too many transfers, max: {0}")]
+	TooManyTransfers(u32),
```

Comment:
This error is "game over" or at least would require making changes to the node code.
I think when creating InherentData we should use truncate from, so it doesn't really happen.

Reply:
If it is IDP programmatic error then I think we should just `.expect` there. If this can be encountered when handling InherentData from other node, then it is ok to be an error (signalling someone attacks).